nixpkgs_review/nix.py

```diff
@@ -172,9 +269,79 @@ def nix_build(
     ] + shlex.split(args)
 
     try:
-        sh(command)
-    except subprocess.CalledProcessError:
-        pass
+        proc = sh(command, stderr=subprocess.PIPE)
+        stderr = proc.stderr
+    except subprocess.CalledProcessError as e:
+        stderr = e.stderr
+
+    # Remove boring 'copying path' lines from stderr
+    stderr = "\n".join(
+        line
+        for line in stderr.splitlines()
+        if not (
+            line.startswith("copying path '")
+            and line.endswith("from 'https://cache.nixos.org'...")
+        )
+    )
+
+    has_failed_dependencies = []
+    has_timeout = {}
+    for line in stderr.splitlines():
+        if "dependencies couldn't be built" in line:
+            has_failed_dependencies.append(
+                next(item for item in line.split() if "/nix/store" in item)
+                .lstrip("'")
+                .rstrip(":'")
+            )
+        if "timed out after" in line:
+            drv = (
+                next(item for item in line.split() if "/nix/store" in item)
+                .lstrip("'")
+                .rstrip("'")
+            )
+            has_timeout[drv] = line
+
+    drv_path_to_attr = {a.drv_path: a for a in attrs}
+    for drv_path in has_timeout.keys():
+        if drv_path in drv_path_to_attr:
+            attr = drv_path_to_attr[drv_path]
+            attr.build_err_msg = has_timeout[drv_path]
+            attr.timed_out = True
+
+    for drv_path in has_failed_dependencies:
+        if drv_path in drv_path_to_attr:
+            attr = drv_path_to_attr[drv_path]
+            attr.build_err_msg = stderr
+
+            # Without inspecting the build graph, let's just guess
+            # that if something failed to build and there were ANY timeouts
+            # that it's probably the case that this guy's failure to build
+            # was probably caused by the timeout.
+            # e.g. https://github.com/NixOS/nixpkgs/pull/114609
+            if len(has_timeout) > 0:
+                attr.timed_out = True
+
+    return attrs
+
+
+def pre_build_filter(attrs: List[Attr]) -> List[Attr]:
+    for cmd in (
+        cmd
+        for cmd in os.environ.get("NIXPKGS_REVIEW_PRE_BUILD_FILTER", "").split(":")
```

Comment:
That ENV needs documentation.

Reply:
This is a potentially area of discussion, @Mic92.

Despite being a small amount of code, for me this is the most important feature. The idea is that I want to run some code to manipulate the set of `attrs` that the rest of the code deals with, to set some as `skipped=True` based on a dynamic calculation that I can control. Basically a "plugin".

I can think of two possible ways to achieve this feature. One is as written, where `nixpkgs-review` connects with these external plugins by environment variable and subprocesses them. A different way to do it that you might prefer would be to use `import.metadata` `entry_points`, which is kind of a python plugin feature. We could agree on the name of the entry point, and then this `pre_build_filter` function would find the plugins by iterating over the matching entry points rather than looking through an environment variable. The big difference here would be that the plugin would execute inside the same process as the main python interpreter.

I am open to either mechanism.

Reply:
I will have a look this evening into this PR. Thanks for leading the review so far @SuperSandro2000 . Due to diverse requirements and languages used in the community a language-independent solution might be preferred.

Reply:
@ryantm is this something that nixpkgs-update could benefit from as well? If not what is missing?

Reply:
I added some documentation to the README.md